File: android-app/app/src/main/java/com/andernet/weather/data/cache/WeatherCacheManager.kt
package com.andernet.weather.data.cache

import android.content.Context
import androidx.compose.runtime.Stable
import com.andernet.weather.data.model.WeatherData
import com.squareup.moshi.Moshi
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.roundToInt

/**
 * Enhanced cache manager for Android with smart invalidation and memory pressure handling
 * Similar to iOS WeatherCacheManager but adapted for Android patterns
 */
@Singleton
class WeatherCacheManager @Inject constructor(
    @ApplicationContext private val context: Context,
    private val moshi: Moshi
) {
    
    // MARK: - Cache Components
    
    private val memoryCache = ConcurrentHashMap<String, CacheEntry>()
    private val cacheMetadata = ConcurrentHashMap<String, CacheMetadata>()
    private val cacheDir = File(context.cacheDir, "weather_enhanced")
    private val metadataFile = File(cacheDir, "metadata.json")
    
    // MARK: - Cache Configuration
    
    private object CacheConfig {
        const val MEMORY_LIMIT_MB = 20 // 20MB in memory
        const val DISK_LIMIT_MB = 100 // 100MB on disk  
        const val MAX_MEMORY_ITEMS = 15
        const val STALE_TIME_MS = 15 * 60 * 1000L // 15 minutes
        const val MAX_AGE_MS = 6 * 60 * 60 * 1000L // 6 hours
    }
    
    // MARK: - Data Classes
    
    @Stable
    private data class CacheEntry(
        val data: WeatherData,
        val timestamp: Long,
        val size: Int
    )
    
    @Stable
    private data class CacheMetadata(
        val key: String,
        val timestamp: Long,
        val size: Long,
        val accessCount: Int,
        val lastAccess: Long,
        val latitude: Double,
        val longitude: Double
    ) {
        val isStale: Boolean
            get() = System.currentTimeMillis() - timestamp > CacheConfig.STALE_TIME_MS
            
        val isExpired: Boolean
            get() = System.currentTimeMillis() - timestamp > CacheConfig.MAX_AGE_MS
            
        val shouldRefresh: Boolean
            get() = isStale // Simplified - could add network type detection
    }
    
    init {
        cacheDir.mkdirs()
        loadMetadata()
        cleanupExpiredEntries()
    }
    
    // MARK: - Public Interface
    
    /**
     * Get weather data from cache
     */
    suspend fun getCachedWeather(latitude: Double, longitude: Double): WeatherData? {
        return withContext(Dispatchers.IO) {
            val key = cacheKey(latitude, longitude)
            
            // Try memory cache first
            memoryCache[key]?.let { entry ->
                updateAccessMetadata(key)
                return@withContext entry.data
            }
            
            // Try disk cache
            val cacheFile = File(cacheDir, "$key.json")
            if (cacheFile.exists()) {
                try {
                    val json = cacheFile.readText()
                    val adapter = moshi.adapter(WeatherData::class.java)
                    val weatherData = adapter.fromJson(json)
                    
                    if (weatherData != null) {
                        // Load back to memory cache if space available
                        if (memoryCache.size < CacheConfig.MAX_MEMORY_ITEMS) {
                            val entry = CacheEntry(
                                data = weatherData,
                                timestamp = System.currentTimeMillis(),
                                size = json.length
                            )
                            memoryCache[key] = entry
                        }
                        
                        updateAccessMetadata(key)
                        return@withContext weatherData
                    }
                } catch (e: Exception) {
                    // Invalid cache file - remove it
                    cacheFile.delete()
                    cacheMetadata.remove(key)
                }
            }
            
            null
        }
    }
    
    /**
     * Store weather data in cache
     */
    suspend fun cacheWeather(
        weatherData: WeatherData,
        latitude: Double,
        longitude: Double
    ) {
        withContext(Dispatchers.IO) {
            val key = cacheKey(latitude, longitude)
            val adapter = moshi.adapter(WeatherData::class.java)
            val json = adapter.toJson(weatherData)
            
            // Store in memory cache
            val entry = CacheEntry(
                data = weatherData,
                timestamp = System.currentTimeMillis(),
                size = json.length
            )
            
            // Manage memory cache size
            if (memoryCache.size >= CacheConfig.MAX_MEMORY_ITEMS) {
                evictLeastRecentlyUsed()
            }
            memoryCache[key] = entry
            
            // Store on disk
            val cacheFile = File(cacheDir, "$key.json")
            try {
                cacheFile.writeText(json)
                
                // Update metadata
                val metadata = CacheMetadata(
                    key = key,
                    timestamp = System.currentTimeMillis(),
                    size = json.length.toLong(),
                    accessCount = 1,
                    lastAccess = System.currentTimeMillis(),
                    latitude = latitude,
                    longitude = longitude
                )
                cacheMetadata[key] = metadata
                saveMetadata()
                
                // Cleanup disk if needed
                cleanupDiskCacheIfNeeded()
            } catch (e: Exception) {
                // Failed to write to disk - at least we have memory cache
                e.printStackTrace()
            }
        }
    }
    
    /**
     * Check if cached data should be refreshed
     */
    fun shouldRefreshCache(latitude: Double, longitude: Double): Boolean {
        val key = cacheKey(latitude, longitude)
        val metadata = cacheMetadata[key] ?: return true
        return metadata.shouldRefresh || metadata.isExpired
    }
    
    /**
     * Warm cache for given locations
     */
    suspend fun warmCache(locations: List<Pair<Double, Double>>) {
        withContext(Dispatchers.IO) {
            locations.forEach { (latitude, longitude) ->
                if (shouldRefreshCache(latitude, longitude)) {
                    // Cache warming would trigger background fetch
                    // Implementation depends on your weather service architecture
                }
            }
        }
    }
    
    /**
     * Clear expired cache entries
     */
    suspend fun clearExpired() {
        withContext(Dispatchers.IO) {
            cleanupExpiredEntries()
        }
    }
    
    /**
     * Get cache statistics
     */
    fun getCacheStats(): CacheStats {
        val totalMemorySize = memoryCache.values.sumOf { it.size }
        val totalDiskSize = cacheMetadata.values.sumOf { it.size }
        
        return CacheStats(
            memoryItems = memoryCache.size,
            diskItems = cacheMetadata.size,
            memorySize = totalMemorySize / (1024 * 1024), // MB
            diskSize = totalDiskSize / (1024 * 1024) // MB
        )
    }
    
    // MARK: - Private Methods
    
    private fun cacheKey(latitude: Double, longitude: Double): String {
        val lat = (latitude * 10000).roundToInt() / 10000.0
        val lon = (longitude * 10000).roundToInt() / 10000.0
        return "weather_${lat}_${lon}"
    }
    
    private fun updateAccessMetadata(key: String) {
        val metadata = cacheMetadata[key] ?: return
        val updated = metadata.copy(
            accessCount = metadata.accessCount + 1,
            lastAccess = System.currentTimeMillis()
        )
        cacheMetadata[key] = updated
        
        // Save metadata periodically
        if (updated.accessCount % 10 == 0) {
            saveMetadata()
        }
    }
    
    private fun evictLeastRecentlyUsed() {
        // Find LRU item in memory cache
        var lruKey: String? = null
        var oldestAccess = Long.MAX_VALUE
        
        for ((key, _) in memoryCache) {
            val metadata = cacheMetadata[key]
            if (metadata != null && metadata.lastAccess < oldestAccess) {
                oldestAccess = metadata.lastAccess
                lruKey = key
            }
        }
        
        lruKey?.let { memoryCache.remove(it) }
    }
    
    private fun cleanupExpiredEntries() {
        val now = System.currentTimeMillis()
        val expiredKeys = cacheMetadata.values.filter { 
            now - it.timestamp > CacheConfig.MAX_AGE_MS 
        }.map { it.key }
        
        expiredKeys.forEach { key ->
            // Remove from memory
            memoryCache.remove(key)
            
            // Remove from disk
            File(cacheDir, "$key.json").delete()
            
            // Remove metadata
            cacheMetadata.remove(key)
        }
        
        if (expiredKeys.isNotEmpty()) {
            saveMetadata()
        }
    }
    
    private fun cleanupDiskCacheIfNeeded() {
        val totalSize = cacheMetadata.values.sumOf { it.size }
        val limitBytes = CacheConfig.DISK_LIMIT_MB * 1024 * 1024L
        
        if (totalSize > limitBytes) {
            // Sort by last access (LRU)
            val sortedEntries = cacheMetadata.values.sortedBy { it.lastAccess }
            
            var currentSize = totalSize
            val targetSize = (limitBytes * 0.8).toLong() // Clean to 80% of limit
            
            for (metadata in sortedEntries) {
                if (currentSize <= targetSize) break
                
                // Remove from all caches
                memoryCache.remove(metadata.key)
                File(cacheDir, "${metadata.key}.json").delete()
                cacheMetadata.remove(metadata.key)
                
                currentSize -= metadata.size
            }
            
            saveMetadata()
        }
    }
    
    private fun loadMetadata() {
        if (metadataFile.exists()) {
            try {
                val json = metadataFile.readText()
                val adapter = moshi.adapter<Map<String, CacheMetadata>>(
                    Map::class.java, 
                    String::class.java, 
                    CacheMetadata::class.java
                )
                val loaded = adapter.fromJson(json)
                if (loaded != null) {
                    cacheMetadata.putAll(loaded)
                }
            } catch (e: Exception) {
                e.printStackTrace()
                cacheMetadata.clear()
            }
        }
    }
    
    private fun saveMetadata() {
        try {
            val adapter = moshi.adapter<Map<String, CacheMetadata>>(
                Map::class.java,
                String::class.java,
                CacheMetadata::class.java
            )
            val json = adapter.toJson(cacheMetadata.toMap())
            metadataFile.writeText(json ?: "{}")
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
    
    // MARK: - Data Classes
    
    @Stable
    data class CacheStats(
        val memoryItems: Int,
        val diskItems: Int,
        val memorySize: Int, // MB
        val diskSize: Int    // MB
    )
}

File: weather/Sources/Managers/WeatherCacheManager.swift
//
//  WeatherCacheManager.swift
//  weather
//
//  Enhanced cache manager with memory pressure handling and smart invalidation
//

import Foundation
import UIKit
import OSLog

/**
 * Enhanced cache manager with memory pressure handling, cache warming, and intelligent invalidation
 */
@MainActor
class WeatherCacheManager {
    static let shared = WeatherCacheManager()
    
    // MARK: - Cache Components
    
    /// Memory cache with automatic eviction under pressure
    private let memoryCache = NSCache<NSString, WeatherData>()
    
    /// Disk cache directory
    private let diskCacheDirectory: URL
    
    /// Cache metadata for smart invalidation
    private var cacheMetadata: [String: CacheMetadata] = [:]
    
    /// Logger for cache operations
    private let logger = Logger(subsystem: "WeatherApp.Cache", category: "CacheManager")
    
    // MARK: - Cache Configuration
    
    private struct CacheConfig {
        static let memoryLimit = 50 * 1024 * 1024 // 50MB in memory
        static let diskLimit = 200 * 1024 * 1024 // 200MB on disk
        static let itemCountLimit = 20 // Max items in memory
        static let staleTime: TimeInterval = 15 * 60 // 15 minutes before data is stale
        static let maxAge: TimeInterval = 6 * 60 * 60 // 6 hours max age
    }
    
    // MARK: - Cache Metadata
    
    private struct CacheMetadata: Codable {
        let key: String
        let timestamp: Date
        let size: Int64
        let accessCount: Int
        let lastAccess: Date
        let location: (latitude: Double, longitude: Double)
        
        var isStale: Bool {
            Date().timeIntervalSince(timestamp) > CacheConfig.staleTime
        }
        
        var isExpired: Bool {
            Date().timeIntervalSince(timestamp) > CacheConfig.maxAge
        }
        
        var shouldRefresh: Bool {
            // Refresh if stale AND user is likely on WiFi (simplified heuristic)
            isStale && ProcessInfo.processInfo.isLowPowerModeEnabled == false
        }
    }
    
    // MARK: - Initialization
    
    private init() {
        // Create disk cache directory
        let cacheDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!
        diskCacheDirectory = cacheDirectory.appendingPathComponent("WeatherCache")
        
        try? FileManager.default.createDirectory(at: diskCacheDirectory, withIntermediateDirectories: true)
        
        setupMemoryCache()
        loadCacheMetadata()
        setupMemoryWarningHandling()
        
        logger.info("WeatherCacheManager initialized with disk cache at: \\(diskCacheDirectory.path)")
    }
    
    private func setupMemoryCache() {
        memoryCache.countLimit = CacheConfig.itemCountLimit
        memoryCache.totalCostLimit = CacheConfig.memoryLimit
        
        // Eviction delegate to update metadata
        memoryCache.delegate = self
    }
    
    private func setupMemoryWarningHandling() {
        NotificationCenter.default.addObserver(
            forName: UIApplication.didReceiveMemoryWarningNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleMemoryWarning()
        }
        
        NotificationCenter.default.addObserver(
            forName: UIApplication.didEnterBackgroundNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            Task {
                await self?.cleanupAndOptimize()
            }
        }
    }
    
    // MARK: - Public Cache Interface
    
    /// Get weather data from cache
    func getCachedWeather(latitude: Double, longitude: Double) -> WeatherData? {
        let key = cacheKey(latitude: latitude, longitude: longitude)
        
        // Try memory cache first
        if let cached = memoryCache.object(forKey: key as NSString) {
            updateAccessMetadata(for: key)
            logger.debug("Cache HIT (memory): \\(key)")
            return cached
        }
        
        // Try disk cache
        let diskURL = diskCacheDirectory.appendingPathComponent(key).appendingPathExtension("json")
        
        do {
            let data = try Data(contentsOf: diskURL)
            let weather = try JSONDecoder().decode(WeatherData.self, from: data)
            
            // Load back to memory cache
            let cost = data.count
            memoryCache.setObject(weather, forKey: key as NSString, cost: cost)
            
            updateAccessMetadata(for: key)
            logger.debug("Cache HIT (disk): \\(key)")
            return weather
        } catch {
            logger.warning("Cache MISS: \\(key) - \\(error.localizedDescription)")
            return nil
        }
    }
    
    /// Store weather data in cache
    func cacheWeather(_ weather: WeatherData, latitude: Double, longitude: Double) {
        let key = cacheKey(latitude: latitude, longitude: longitude)
        
        Task {
            do {
                let data = try JSONEncoder().encode(weather)
                let cost = data.count
                
                // Store in memory cache
                memoryCache.setObject(weather, forKey: key as NSString, cost: cost)
                
                // Store on disk
                let diskURL = diskCacheDirectory.appendingPathComponent(key).appendingPathExtension("json")
                try data.write(to: diskURL)
                
                // Update metadata
                let metadata = CacheMetadata(
                    key: key,
                    timestamp: Date(),
                    size: Int64(data.count),
                    accessCount: 1,
                    lastAccess: Date(),
                    location: (latitude: latitude, longitude: longitude)
                )
                cacheMetadata[key] = metadata
                saveCacheMetadata()
                
                logger.info("Cached weather data: \\(key) (\\(ByteCountFormatter.string(fromByteCount: Int64(cost), countStyle: .binary)))")
                
                // Cleanup if over disk limit
                await cleanupDiskCacheIfNeeded()
            } catch {
                logger.error("Failed to cache weather data: \\(error)")
            }
        }
    }
    
    /// Check if cached data should be refreshed
    func shouldRefreshCache(latitude: Double, longitude: Double) -> Bool {
        let key = cacheKey(latitude: latitude, longitude: longitude)
        guard let metadata = cacheMetadata[key] else { return true }
        
        return metadata.shouldRefresh || metadata.isExpired
    }
    
    /// Preload cache for given locations (cache warming)
    func warmCache(for locations: [(latitude: Double, longitude: Double)]) {
        Task(priority: .background) {
            for location in locations {
                // Only warm cache if we don't have recent data
                if shouldRefreshCache(latitude: location.latitude, longitude: location.longitude) {
                    logger.info("Cache warming for location: (\\(location.latitude), \\(location.longitude))")
                    // Note: This would typically trigger a background fetch
                    // Implementation depends on your weather service architecture
                }
            }
        }
    }
    
    // MARK: - Cache Management
    
    private func handleMemoryWarning() {
        logger.warning("Memory warning received - clearing memory cache")
        memoryCache.removeAllObjects()
        
        // Keep only most recently accessed items metadata
        let sortedMetadata = cacheMetadata.values.sorted { $0.lastAccess > $1.lastAccess }
        let keepCount = min(5, sortedMetadata.count)
        
        cacheMetadata = Dictionary(uniqueKeysWithValues: 
            sortedMetadata.prefix(keepCount).map { ($0.key, $0) }
        )
        
        saveCacheMetadata()
    }
    
    private func cleanupAndOptimize() async {
        logger.info("Running cache cleanup and optimization")
        
        // Remove expired entries
        let now = Date()
        var expiredKeys: [String] = []
        
        for (key, metadata) in cacheMetadata {
            if metadata.isExpired {
                expiredKeys.append(key)
            }
        }
        
        // Remove expired files and metadata
        for key in expiredKeys {
            let diskURL = diskCacheDirectory.appendingPathComponent(key).appendingPathExtension("json")
            try? FileManager.default.removeItem(at: diskURL)
            cacheMetadata.removeValue(forKey: key)
        }
        
        if !expiredKeys.isEmpty {
            logger.info("Cleaned up \\(expiredKeys.count) expired cache entries")
            saveCacheMetadata()
        }
        
        await cleanupDiskCacheIfNeeded()
    }
    
    private func cleanupDiskCacheIfNeeded() async {
        let totalSize = cacheMetadata.values.reduce(0) { $0 + $1.size }
        
        if totalSize > CacheConfig.diskLimit {
            logger.warning("Disk cache over limit (\\(ByteCountFormatter.string(fromByteCount: totalSize, countStyle: .binary))) - starting cleanup")
            
            // Remove least recently used items
            let sortedMetadata = cacheMetadata.values.sorted { $0.lastAccess < $1.lastAccess }
            
            var currentSize = totalSize
            var removedCount = 0
            
            for metadata in sortedMetadata {
                if currentSize <= Int64(Double(CacheConfig.diskLimit) * 0.8) { // Clean to 80% of limit
                    break
                }
                
                let diskURL = diskCacheDirectory.appendingPathComponent(metadata.key).appendingPathExtension("json")
                try? FileManager.default.removeItem(at: diskURL)
                cacheMetadata.removeValue(forKey: metadata.key)
                
                currentSize -= metadata.size
                removedCount += 1
            }
            
            logger.info("Cleaned up \\(removedCount) cache entries to free space")
            saveCacheMetadata()
        }
    }
    
    // MARK: - Metadata Management
    
    private func updateAccessMetadata(for key: String) {
        guard var metadata = cacheMetadata[key] else { return }
        
        metadata = CacheMetadata(
            key: metadata.key,
            timestamp: metadata.timestamp,
            size: metadata.size,
            accessCount: metadata.accessCount + 1,
            lastAccess: Date(),
            location: metadata.location
        )
        
        cacheMetadata[key] = metadata
        
        // Save metadata periodically (every 10 accesses)
        if metadata.accessCount % 10 == 0 {
            saveCacheMetadata()
        }
    }
    
    private func loadCacheMetadata() {
        let metadataURL = diskCacheDirectory.appendingPathComponent("metadata.json")
        
        do {
            let data = try Data(contentsOf: metadataURL)
            cacheMetadata = try JSONDecoder().decode([String: CacheMetadata].self, from: data)
            logger.debug("Loaded cache metadata for \\(cacheMetadata.count) entries")
        } catch {
            logger.info("No existing cache metadata found - starting fresh")
            cacheMetadata = [:]
        }
    }
    
    private func saveCacheMetadata() {
        let metadataURL = diskCacheDirectory.appendingPathComponent("metadata.json")
        
        do {
            let data = try JSONEncoder().encode(cacheMetadata)
            try data.write(to: metadataURL)
        } catch {
            logger.error("Failed to save cache metadata: \\(error)")
        }
    }
    
    // MARK: - Utilities
    
    private func cacheKey(latitude: Double, longitude: Double) -> String {
        // Round to reasonable precision for cache key
        let lat = String(format: "%.4f", latitude)
        let lon = String(format: "%.4f", longitude)
        return "weather_\\(lat)_\\(lon)"
    }
    
    // MARK: - Statistics
    
    func getCacheStatistics() -> (memoryItems: Int, diskItems: Int, totalSize: String) {
        let memoryItems = memoryCache.countLimit
        let diskItems = cacheMetadata.count
        let totalSize = cacheMetadata.values.reduce(0) { $0 + $1.size }
        
        return (
            memoryItems: memoryItems,
            diskItems: diskItems,
            totalSize: ByteCountFormatter.string(fromByteCount: totalSize, countStyle: .binary)
        )
    }
}

// MARK: - NSCacheDelegate

extension WeatherCacheManager: NSCacheDelegate {
    nonisolated func cache(_ cache: NSCache<AnyObject, AnyObject>, willEvictObject obj: AnyObject) {
        // Log eviction for debugging
        Task { @MainActor in
            logger.debug("Memory cache evicting object due to pressure")
        }
    }
}

// MARK: - Cache Metadata Extensions

private extension WeatherCacheManager.CacheMetadata {
    enum CodingKeys: String, CodingKey {
        case key, timestamp, size, accessCount, lastAccess
        case latitude, longitude
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        key = try container.decode(String.self, forKey: .key)
        timestamp = try container.decode(Date.self, forKey: .timestamp)
        size = try container.decode(Int64.self, forKey: .size)
        accessCount = try container.decode(Int.self, forKey: .accessCount)
        lastAccess = try container.decode(Date.self, forKey: .lastAccess)
        
        let latitude = try container.decode(Double.self, forKey: .latitude)
        let longitude = try container.decode(Double.self, forKey: .longitude)
        location = (latitude: latitude, longitude: longitude)
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(key, forKey: .key)
        try container.encode(timestamp, forKey: .timestamp)
        try container.encode(size, forKey: .size)
        try container.encode(accessCount, forKey: .accessCount)
        try container.encode(lastAccess, forKey: .lastAccess)
        try container.encode(location.latitude, forKey: .latitude)
        try container.encode(location.longitude, forKey: .longitude)
    }
}